### PR TITLE
Mtw 21

### LINKED ISSUE #21: Create Outcomes Screen
### Ticket
As a youth, I want to see all my tasks for each Focus Area listed under its corresponding outcome, so that I know exactly what tasks need to be done in order to complete each outcome.

### Acceptance Criteria 
Given I've tapped on a Focus Area from the Pod screen, when the Outcomes Screen launches, then I see that I'm on one of the 3 Focus Areas' Outcomes Screen, which contains all the tasks listed under their corresponding outcomes for that Focus Area.

### PR DESCRIPTION
**Pull request by ryanbeckwith**
# Overview

The frontend components for a generalizable Outcomes page are included in this PR, as well as backend endpoints to dynamically update Task and Outcome components based on Salesforce data. Frontend infrastructure for querying such endpoints is included as well. Furthermore, this PR includes code for updating Salesforce checkbox fields when the user checks off tasks on the frontend.

In the future, we plan to utilize the authorization token framework to obtain individualized user data so that the values of each checkbox are specific to each user. Currently, our testing only involves Trainee Pod TR-1 and Youth FakeD.

If there are any other questions about the content of this PR, feel free to ask either myself, Julia, or Alec!

# Steps to Test this PR:

1. Load the Expo app on a mobile device.
2. Manually verify that frontend interaction successfully updates Salesforce for example user data.

# Standard Checklist:

- [x] Does the code work? Does it perform its intended function, the logic is correct etc.
- [x] Have you merged the most recent version of staging into your branch?
- [x] Are all merge conflicts resolved? 
- [x] Can any logging or debugging code be removed? (Any debugging statements should be removed)
- [x] Is the `diff` the **minimum amount** needed to achieve this ticket's goal?

### CODE REVIEW

File: frontend/components/outcomes/OutcomesScreen.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                data={dataTemp}
+                renderItem={({ item }) => {
+                    return (
+                        <Outcome data={[item]}/>
+                    );
+                }}
+                keyExtractor={item => item.title}
+            />
+        </SafeAreaView>
+    );
```
Comment (titapa-punpun) [inline frontend/components/outcomes/OutcomesScreen.js]: Can you add a valid default state that shows if it errors or if there's no data to fetch? 
Default Error Message can be: "There are no outcomes to display. Please contact your manager or More Than Words administrator if you think this is an error."

File: frontend/components/outcomes/Outcome.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                    checked: true
+                },
+                {
+                    key: "Identify a Post MTW Plan",
+                    starIsFilled: false,
+                    checked: true
+                },
+            ]
+        },
+    ];
```
Comment (titapa-punpun) [inline frontend/components/outcomes/Outcome.js]: I think the `defaultData` should be empty and if there is no data to fetch, this component should not render and the app should display an error message (see other comment).

File: frontend/components/outcomes/Task.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                    color={this.state.starIsFilled ? "#C4C4C4" : "#3F3F3F"}
+                    backgroundColor='transparent'
+                    underlayColor='transparent'
+                    onPress={this.state.starIsFilled ? null : () => {
+                        this.updateSalesforce();
+                        this.setState({checked: !this.state.checked});
+                    }}
+                />
+            </View>
+        );
```
Comment (titapa-punpun) [inline frontend/components/outcomes/Task.js]: If YDM approved the task, it should be greyed out, the youth checkbox should be disabled, and the star should be disabled as well. Can't see why youth would want to star a task that's already been approved.

File: frontend/components/outcomes/Task.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    async updateSalesforce() {
+        let updated_value = !this.state.checked
+        if (updated_value === true) {
+            updated_value = "True"
+        } else {
+            updated_value = "False"
+        }
+        axios.get(server_add + "/updateCheckbox", {
```
Comment (titapa-punpun) [inline frontend/components/outcomes/Task.js]: There should be error handling for this endpoint -- what if it fails?

File: frontend/components/outcomes/Task.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            checkboxContainer: {
+                paddingRight: 0,
+            }
+        });
+    }
+
+    async updateSalesforce() {
+        let updated_value = !this.state.checked
+        if (updated_value === true) {
+            updated_value = "True"
```
Comment (titapa-punpun) [inline frontend/components/outcomes/Task.js]: Can reuse bool:
`updated_value = !this.state.checked`

UPDATE: see Slack channel

File: frontend/components/outcomes/OutcomesHome.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
+/*
```
Comment (titapa-punpun) [inline frontend/components/outcomes/OutcomesHome.js]: Is this file a duplicate of `OutcomesScreen.js`? I couldn't find any usages of this component.

File: frontend/components/outcomes/Task.js [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    render() {
+        let name = this.props.name;
+
+        return (
+            <View style={this.styles().taskContainer}>
+                <Icon
+                    name={this.state.starIsFilled ? "star" : "star-border"}
+                    color={this.state.starIsFilled ? "#C4C4C4" : "#FF4646"}
```
Comment (titapa-punpun) [inline frontend/components/outcomes/Task.js]: For now, can you hardcode this in as 
`name={false ? "star" : "star-border"}`
`color={false ? "#C4C4C4" : "#FF4646"}`

later, when you have another variable for star filling, you can replace the boolean constants with the variable

File: frontend/components/outcomes/Task.js [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+                <Text style={this.styles().text}>
+                    {name}
+                </Text>
+
+                <Icon.Button
+                    style={this.styles().checkboxContainer}
+                    name={this.state.checked ? 'check-box' : 
+                          'check-box-outline-blank'}
+                    iconStyle={this.styles().checkbox}
+                    color={this.state.starIsFilled ? "#C4C4C4" : "#3F3F3F"}
```
Comment (titapa-punpun) [inline frontend/components/outcomes/Task.js]: Can you rename `starIsFilled` to `checkboxFilled` or `youthCheckboxFilled`?

**Review by titapa-punpun — CHANGES_REQUESTED** [review]
Solid PR!! Can you delete `package.json`, `webpack.config.js`, and `yarn.lock` in the root directory? They should only be in the `frontend` folder. Other than that, please see comments in the individual files - thanks!

File: backend/app.py [hunk 9]
```diff
@@ -15,6 +15,41 @@
...
+
+@app.route("/updateCheckbox")
+def updateSalesforce():
+    tr_pod_id = request.args.get('tr_pod_id')
+    task_title = request.args.get('task_title')
+    new_value = request.args.get('new_value')
+    if (new_value == 'True'):
+        new_value = True
+    else:
+        new_value = False
```
Comment (titapa-punpun) [inline backend/app.py]: This can probably be deleted if `new_value` passed in from frontend is a `bool` instead of a `string`

**Comment by ryanbeckwith** [comment]
As of 3/2/21, the following changes have been implemented:
-  A valid default state, along with an error message, for any endpoint request that failed to obtain outcomes.
- YDM approval now greys out the star and removes the ability to click on it, and restores the star to unfilled.
- Separate variables (with better names) were added for the filled star and YDM approval attributes.
- Backend GET requests for retrieving Outcome and Task data from Salesforce have been generalized across all pods (and works with high reliability).
- Backend POST requests for the checkbox endpoint have been generalized across all pods (and work with varying reliability)
- Authentication has been enabled for the two relevant backend endpoints (potential varying reliability due to 401 issue).
- User data is derived from the current user logged into the app, not a default user.
- Star and checkbox state is locally preserved during interaction with the Outcome screen, which fixed a bug where the state was reverted to the initial Salesforce value on screen load if the accordion was minimized and reopened.

Furthermore, these are the known issues that still need to be fixed:
- 401 authorization issue occurs for multiple consecutive requests (either GET or POST) to either relevant endpoint if the requests are made within a short period of time.
- 401 authorization issue for updating specific checkbox values in the Associate pod, but not the Trainee pod.
- Undefined error when clicking the star of one specific task (the star in the "Complete Getting Started Module" task of the Life Essentials/Support Network Outcome)

File: startBackendWindows.sh [hunk 10]
```diff
@@ -3,4 +3,5 @@ cd backend
 py -m venv backendEnv
 source backendEnv/scripts/activate
 pip install -r requirements.txt
-eval $(egrep -v '^#' ../.env | xargs) py app.py
\ No newline at end of file
+# eval $(egrep -v '^#' ../.env | xargs) py app.py
+py ~/OneDrive/Documents/Ryan/School/"College 2nd Year"/JumboCode/MoreThanWords/MoreThanWords/backend/app.py
```
Comment (jchin01) [inline startBackendWindows.sh]: Please remove your computer specific path and replace it with the original so anyone is able to run the backend windows script

File: frontend/components/pod_components/HomeScreen.js [hunk 11]
```diff
@@ -42,12 +42,14 @@ export default class HomeScreen extends React.Component{
...
-        return(
+        return (
         <ScrollView style={styles.scrollView}>
             <SafeAreaView style={styles.container}>
             <TouchableOpacity 
                 style={styles.block} 
-                onPress={() => this.props.navigation.navigate('Trainee Pod')}
+                onPress={() => this.props.navigation.navigate('Trainee Pod', {
+                    pod: 'Trainee'
+                })}
```
Comment (jchin01) [inline frontend/components/pod_components/HomeScreen.js:52]: Looks Good!

File: frontend/components/pod_components/AssociatePod.js [hunk 12]
```diff
@@ -13,7 +13,14 @@ export default function AssociatePodScreen({ navigation }) {
...
                 style={styles.block} 
-                onPress={() => navigation.navigate('Random Screen')}
+                onPress={() => {
+                    const { pod } = props.route.params;
+                    props.navigation.navigate('Outcomes', {
+                        pod: pod,
+                        focus_area: "COM",
+                        title: "Competencies"
+                    });
+                }}
```
Comment (jchin01) [inline frontend/components/pod_components/AssociatePod.js:23]: These look good for now to get basic functionality! Nice we can finally start to see outcomes for different pods

File: frontend/components/pod_components/PartnerPod.js [hunk 13]
```diff
@@ -36,7 +36,7 @@ export default function PartnerPodScreen({ navigation }) {
         
             <TouchableOpacity 
                 style={styles.block} 
-                onPress={() => navigation.navigate('Random Screen')}
+                onPress={() => navigation.navigate('Outcomes')}
```
Comment (jchin01) [inline frontend/components/pod_components/PartnerPod.js]: These lines in the partner pod cause the app to error. I think you need to pass in the information to this pod the same way you did for associates and trainee

File: backend/app.py [hunk 14]
```diff
@@ -7,14 +7,75 @@
...
+    # Format response
+    response = {}
+    for name_and_label in field_names_and_labels:
+        response[name_and_label[0]] = {
+            "name": name_and_label[1],
+            "value": None
+        }
+    for name, value in sf_result["records"][0].items():
+        if name in response.keys():
+            response[name]["value"] = value
```
Comment (jchin01) [inline backend/app.py:49]: Nice job cleaning up the response thats sent back to the frontend

File: backend/app.py [hunk 15]
```diff
@@ -94,4 +155,4 @@ def outcomes(user):
     return sf_result
 
 if __name__ == '__main__':
-    app.run(debug=True, host='0.0.0.0')
+    app.run(debug=True, host=os.environ['LOCAL_IPV4'])
```
Comment (jchin01) [inline backend/app.py]: Please change this back to 0.0.0.0

File: frontend/components/outcomes/Outcome.js [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+                        style={styles.contentStyle}> 
+                        {itemComponents}
+                    </View>
+                );
+            }}
+        />
+    );
+};
+
+export default Outcome;
```
Comment (jchin01) [inline frontend/components/outcomes/Outcome.js:90]: I like the modularity of breaking the outcome into its own component

**Review by jchin01 — CHANGES_REQUESTED** [review]
Just a quick fix of some small things and your PR should be ready!

**Review by jchin01 — APPROVED** [review]

**Review by titapa-punpun — APPROVED** [review]

**Pull request merged**